contracts/TxRelay.sol

```diff
@@ -22,21 +25,24 @@ contract TxRelay {
         bytes32 sigR,
         bytes32 sigS,
         address destination,
-        bytes data
+        bytes data,
+        uint listNum
     ) public {
 
+        // only allow senders from the whitelist specified by the user,
+        // 0 means no whitelist.
+        require(listNum == 0 || whitelist[listNum][msg.sender]);
+
         address claimedSender = getAddress(data);
-        // relay :: nonce :: destination :: data :: relayer
-        bytes32 h = keccak256(this, nonce[claimedSender], destination, data);
+        // relay :: whitelist :: nonce :: destination :: data
+        bytes32 h = keccak256(this, listNum, nonce[claimedSender], destination, data);
```

Comment:
Should we be doing EIP 191 here since we’re already changing it?

Reply:
Right, that is addressed in another PR. Will rebase on that once it's merged into `release/2.0.0`